File: day_03/src/main.rs
use std::io::{BufRead, BufReader};
use std::fs::File;
    
fn main() {
    let tests = [
        [
            vec!["R8","U5","L5","D3"],
            vec!["U7","R6","D4","L4"],
            vec!["6"]
        ],
        [
            vec!["R75","D30","R83","U83","L12","D49","R71","U7","L72"],
            vec!["U62","R66","U55","R34","D71","R55","D58","R83"],
            vec!["159"],
        ],
        [
            vec!["R98","U47","R26","D63","R33","U87","L62","D20","R33","U53","R51"],
            vec!["U98","R91","D20","R16","D67","R40","U7","R15","U6","R7"],
            vec!["135"]
        ],
    ];

    for test in tests.iter() {
        println!("Test\n {:?} {:?}\nshould equal\n {:?}", test[0], test[1], test[2]);
        println!("Actual:\n {:?}\n", closest_intersection(&test[0], &test[1]));
    }

    let input = get_input();
    println!("Final answer part 1: {}", closest_intersection(&input[0], &input[1]));
    
}

fn get_input() -> [Vec<&'static str>; 2] {
    let f = File::open("input/input.txt").unwrap();
    let reader = BufReader::new(f);
    let mut values1 = Vec::new();
    let mut values2 = Vec::new();
    
    for (index,line) in reader.lines().enumerate() {
        if index == 0 {
            values1 = line.unwrap().split(",").collect();
        }
        else {
            values2 = line.unwrap().split(",").collect();
        }
    }
    println!("{:?}",values1);
    return [values1, values2];
}

fn closest_intersection(wire_1: &Vec<&str>, wire_2: &Vec<&str>) -> i32 {
    let path_1 = calculate_path(wire_1);
    let path_2 = calculate_path(wire_2);
    let intersections = get_intersections(path_1, path_2);
    return get_min_distance(intersections);
}

fn calculate_path(wire: &Vec<&str>) -> Vec<[i32;2]> {
    let mut path = Vec::new();
    let mut x = 0;
    let mut y = 0;
    path.push([x,y]);
    for step in wire.iter() {
        let direction = step.chars().nth(0).unwrap();
        let distance: i32 = step[1..].parse().unwrap();
        
        match direction {
            'U' => {
                for yy in (y+1)..=(y+distance) {
                    y = yy;    
                    path.push([x,y]);
                }
            },
            'D' => {
                for yy in ((y-distance)..y).rev() {
                    y = yy;
                    path.push([x,y]);
                }
            },
            'L' => {
                for xx in ((x-distance)..x).rev() {
                    x = xx;
                    path.push([x,y]);
                }
            },
            'R' => {
                for xx in (x+1)..=(x+distance) {
                    x = xx;
                    path.push([x,y]);
                }
            },
            _ => {
                panic!(":c");
            }

        }
    }
    println!("{:?}",path);
    return path;
}

fn get_intersections(path_1: Vec<[i32;2]>, path_2: Vec<[i32;2]>) -> Vec<[i32;2]> {
    let mut intersections = Vec::new();
    for point in path_1.iter() {
        if path_2.contains(point) {
            intersections.push(point.clone());
        }
    }
    for point in path_2.iter() {
        if (path_1.contains(point) && !intersections.contains(point)) {
            intersections.push(point.clone());
        }
    }
    println!("{:?}",intersections);
    return intersections;
}

fn get_min_distance(intersections: Vec<[i32;2]>) -> i32 {
    let mut min = std::i32::MAX;
    for point in intersections.iter() {
        if point[0] == 0 && point[1] == 0 {
            continue;
        }
        let dist = get_manattan_distance([0,0], *point);
        println!("{:?} {}", point, dist);
        if  dist < min {
            min = dist; 
        }
    }
    return min;
}

fn get_manattan_distance(p1: [i32;2], p2: [i32;2]) -> i32 {
    return (p1[0] - p2[0]).abs() + (p1[1] - p2[1]).abs();
}

File: day_03/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type point struct {
	x int
	y int
}

type pathStep struct {
	pos   point
	delay int
}

func main() {
	tests := []([]([]string)){
		{
			[]string{"R8", "U5", "L5", "D3"},
			[]string{"U7", "R6", "D4", "L4"},
			[]string{"6"},
			[]string{"30"},
		},
		{
			[]string{"R75", "D30", "R83", "U83", "L12", "D49", "R71", "U7", "L72"},
			[]string{"U62", "R66", "U55", "R34", "D71", "R55", "D58", "R83"},
			[]string{"159"},
			[]string{"610"},
		},
		{
			[]string{"R98", "U47", "R26", "D63", "R33", "U87", "L62", "D20", "R33", "U53", "R51"},
			[]string{"U98", "R91", "D20", "R16", "D67", "R40", "U7", "R15", "U6", "R7"},
			[]string{"135"},
			[]string{"410"},
		},
	}

	for _, test := range tests {
		closest, leastDelay := closestIntersection(test[0], test[1])
		fmt.Printf("Test: \n%v\n %v\nExpects: \n%v\nActual: \n%v\n", test[0], test[1], test[2], closest)
		fmt.Printf("Pt 2 expects: \n%v\nActual: \n%v\n", test[3], leastDelay)
	}

	input, _ := getInput()
	// fmt.Printf("input: %v\n%v", len(input), input)
	closest, leastDelay := closestIntersection(input[0], input[1])
	fmt.Printf("Final answer, part 1: %v\n", closest)
	fmt.Printf("Final asnwer, part 2: %v\n", leastDelay)

}

func getInput() ([]([]string), error) {
	file, err := os.Open("input/input.txt")
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := bufio.NewReader(file)
	lines := make([]([]string), 0)
	for {
		fileLine, err := reader.ReadString('\n')

		line := make([]string, 0)
		for _, item := range strings.Split(fileLine, ",") {
			line = append(line, strings.Trim(item, "\"\n"))
		}
		lines = append(lines, line)

		if err != nil {
			break
		}
	}

	return lines, nil
}

func closestIntersection(wire1 []string, wire2 []string) (int, int) {
	path1 := calculatePath(wire1)
	path2 := calculatePath(wire2)
	intersections := getIntersections(path1, path2)
	return getMinDistance(intersections), getMinDelay(intersections)
}

func calculatePath(wire []string) []pathStep {
	path := make([]pathStep, 0)
	x := 0
	y := 0
	delay := 0
	path = append(path, pathStep{point{x, y}, delay})

	for _, step := range wire {
		direction := step[0]
		distance, _ := strconv.Atoi(step[1:])

		switch direction {
		case 'U':
			limit := y + distance
			for yy := y + 1; yy <= limit; yy++ {
				y = yy
				pos := point{x, y}
				// delay = getDelay(delay, path, pos)
				delay++
				path = append(path, pathStep{pos, delay})
			}
		case 'D':
			limit := y - distance
			for yy := y - 1; yy >= limit; yy-- {
				y = yy
				pos := point{x, y}
				// delay = getDelay(delay, path, pos)
				delay++
				path = append(path, pathStep{pos, delay})
			}
		case 'L':
			limit := x - distance
			for xx := x - 1; xx >= limit; xx-- {
				x = xx
				pos := point{x, y}
				// delay = getDelay(delay, path, pos)
				delay++
				path = append(path, pathStep{pos, delay})
			}
		case 'R':
			limit := x + distance
			for xx := x + 1; xx <= limit; xx++ {
				x = xx
				pos := point{x, y}
				// delay = getDelay(delay, path, pos)
				delay++
				path = append(path, pathStep{pos, delay})
			}
		}

	}
	// fmt.Printf("getPath: %v\n", path)
	return path
}

// func getDelay(delay int, path []pathStep, newPos point) int {
// 	for _, s := range path {
// 		if s.pos.x == newPos.x && s.pos.y == newPos.y {
// 			return s.delay
// 		}
// 	}

// 	return delay + 1
// }

func getIntersections(path1 []pathStep, path2 []pathStep) []pathStep {
	intersections := make([]pathStep, 0)
	for _, s := range path1 {
		if s.pos.x == 0 && s.pos.y == 0 {
			continue
		}
		for _, s2 := range path2 {
			if s.pos.x == 0 && s.pos.y == 0 {
				continue
			}
			if s.pos.x == s2.pos.x && s.pos.y == s2.pos.y {
				intersections = append(intersections, pathStep{point{s.pos.x, s.pos.y}, s.delay + s2.delay})
			}
		}
	}
	// fmt.Printf("getIntersections: %v\n", intersections)
	return intersections
}

func getMinDistance(path []pathStep) int {
	min := math.MaxInt64
	for _, s := range path {
		d := manhattanDistance(s.pos, point{0, 0})
		if d < min {
			min = d
		}
	}
	return min
}

func manhattanDistance(point1 point, point2 point) int {
	return abs(point1.x-point2.x) + abs(point1.y-point2.y)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func getMinDelay(path []pathStep) int {
	min := math.MaxInt64
	for _, s := range path {
		if s.delay < min {
			min = s.delay
		}
	}

	return min
}
